text/287-cli-deprecation-warnings.md

```diff
@@ -0,0 +1,190 @@
+  so this feature might not provide much value.
+2. We risk overloading customers with warnings,
+  training them to ignore our output if there's too much of it.
+  A warning must be read in order to be effective, after all.
+
+### What changes are required to enable this change?
+
+One of the core tenets behind this RFC is that we want to avoid the need to add any code to the deprecated elements manually.
+Just deprecating the element should automatically trigger the warnings,
+without requiring any other code changes to the element being deprecated.
+
+There are two high-level components of this change:
+
+#### 1. Communication between the framework and the CLI
+
+Since we want to show the warnings when the CLI invokes synthesis,
+we need to communicate between the framework (where those deprecated usages happen)
+and the CLI (where the warnings will be shown).
+The natural way to do that is through the Cloud Assembly.
+
+We already have a facility for adding messages to the construct tree's metadata
+(through the Node class in the `constructs` library),
+which are then [collected into the Stack Cloud Assembly artifact type](https://github.com/aws/aws-cdk/blob/34a921b9667402b6d90731f1fd9e3de1ef27f8bf/packages/%40aws-cdk/core/lib/stack-synthesizers/_shared.ts#L63-L101),
+which the [CLI later renders](https://github.com/aws/aws-cdk/blob/34a921b9667402b6d90731f1fd9e3de1ef27f8bf/packages/aws-cdk/lib/api/cxapp/cloud-assembly.ts#L175-L216).
+We should in all likelihood re-use that mechanism.
+
+Note, however, that that requires access to the construct tree,
+which we don't necessarily have in a trivial way from all contexts
+(for example, for an access to a deprecated property of a struct interface).
+Which means we will need to add some mechanism to `core`
+that allows elements to register warnings with without necessarily having access to a construct,
+and which then will be gathered by synthesis and added to the resulting Cloud Assembly
+(parented below a certain construct, perhaps the `App`?).
+This also means we most likely won't be able to link the usage of the deprecated element to a particular `Stack`
+that it was used in.
+
+#### 2. Adding metadata when using deprecated elements
+
+The second element of this change is to actually add the appropriate metadata when any deprecated element is used.
+
+Before we dive into the possible solutions,
+I think it's worth it to talk about what solutions we _don't_ want,
+as that will strongly inform how we approach the problem.
+
+I think we can rule out any solution that requires parsing the customer's source code in search for usages of deprecated elements.
+The downsides of this approach include:
+
+* We would have to write a separate parser for each language supported by the CDK.
+* This additional parsing could have an adverse impact on the performance of CDK commands.
+
+If we agree that we don't want the above solution,
```

Comment:
Let's think of things we wouldn't be able to analyze using dynamic analysis, then we can decide.

------

If someone puts this in their code:

```ts
new Blabla(this, 'bla', {
  bla: undefined,
});
```

we'd have a hard time detecting that. `if (props.bla)` wouldn't detect it. I think `if ('bla' in props)` would, but the latter wouldn't work for jsii languages.

------

If we internally call a deprecated method, we wouldn't be able to distinguish that from a deprecated method called by the user. *Ideally* we'd only warn about the latter.

I guess it shouldn't happen too much in practice, so it wouldn't matter... ?

------

We wouldn't be able to detect usage of deprecated enum values (or I suppose we could--it would be a bunch of additional work though).

------

We wouldn't be able to detect the use of deprecated type declarations (any of enum type, interface type, class type). For example:

```ts
DeprecatedClass obj = new NonDeprecatedClass();
// or
DeprecatedClass obj = construct.giveMeSomething();
```

----

Now this might all be acceptable (because we can argue they'd be rare, and a *correct* solution is fantastically expensive and might not be worth it). But let's trade this off with eyes open.

Reply:
Fair enough. Added these considerations.